Git integration

hpwxf opened the pull request:
Add git integration following a behaviour very close to what `exa` do.

It adds a `--git` option and a block `git` which is only display when `--long` option is enable.

`git` is called only once for the current tree (if the current directory is a subdirectory of git, the status is displayed but if you use a deeper display, e.g. using `--tree`, the git status of repositories in subdirectories is not done). 

---
- [x] Use `cargo fmt`
- [x] Add necessary tests
- [x] Add changelog entry
- [x] Update default config/theme in README (if applicable)
- [x] Update man page at lsd/doc/lsd.md (if applicable)

hpwxf commented:
this PR follows a discussion in https://github.com/Peltoche/lsd/issues/7#issuecomment-1453364120

hpwxf commented:
`git` is enabled by default (when available). The feature `no-git` may be used to force disable it.

The only unsupported target is `i686-pc-windows-gnu` (compilation failure during the link; ex: https://github.com/hpwxf/lsd/actions/runs/4271505990/jobs/7436019267)

hpwxf commented:
Updated with master

zwpaper reviewed (changes requested):
I have done some of the reviews, and allow me to leave the rest to a next time.

I have checked the simple part, it only has some nit requests to change, and I am trying to walk through the complicated parts, trying my best to get this merged

hpwxf commented on src/app.rs:
```diff
@@ -96,21 +96,24 @@ pub struct Cli {
-        value_name = "TYPE",
-        value_parser = ["size", "time", "version", "extension", "none"],
-        overrides_with_all = ["timesort", "sizesort", "extensionsort", "versionsort", "no_sort"]
+    long,
```
I 100% agree. Done.

hpwxf commented:
Thx @zwpaper for the review. I'm sorry for the format issues; my conclusion: don't blindly trust code formatting

hpwxf commented:
@zwpaper is everything ok for you?

zwpaper commented on src/git.rs:
```diff
@@ -0,0 +1,464 @@
+impl GitCache {
+    pub fn new(path: &Path) -> GitCache {
+        let _cachedir = std::fs::canonicalize(path).unwrap();
+        // log::info!("Trying to retrieve Git statuses for {:?}", _cachedir);
```
please drop those debug log comments

hpwxf replied:
Dropped.
`_cachedir` has been also removed since only used as log

zwpaper commented on src/git.rs:
```diff
@@ -0,0 +1,464 @@
+#[cfg(not(feature = "no-git"))]
+impl GitCache {
+    pub fn new(path: &Path) -> GitCache {
+        let _cachedir = std::fs::canonicalize(path).unwrap();
```
log error by https://github.com/lsd-rs/lsd/blob/8c62e9ed12ccf7a737bd7fc24ccfb0b36cf7b7da/src/main.rs#L66 and return empty instead of unwrap directly

hpwxf replied:
removed line. Only used as log.

zwpaper commented on src/git.rs:
```diff
@@ -0,0 +1,464 @@
+
+        let repo = match git2::Repository::discover(path) {
+            Ok(r) => r,
+            Err(_e) => {
```
`Err(e)` and log by https://github.com/lsd-rs/lsd/blob/8c62e9ed12ccf7a737bd7fc24ccfb0b36cf7b7da/src/main.rs#L66

hpwxf replied:
The `Err` case is not a real error. Git2 cannot retrieve git status info. Usually it means that the directory is not a git directory.

zwpaper commented on src/git.rs:
```diff
@@ -0,0 +1,464 @@
+                        statuses.push(elem);
+                    }
+                }
+                Err(_e) => {
```
log error

hpwxf replied:
Done.

zwpaper commented on src/meta/git_file_status.rs:
```diff
@@ -0,0 +1,71 @@
+        colors: &crate::color::Colors,
+        git_symbols: &crate::git_symbol::GitSymbols,
+    ) -> crate::color::ColoredString {
+        use crate::color::{ColoredString, Colors};
```
move `use` and `crate` to the top of the file

hpwxf replied:
Done for all in this file.

zwpaper commented on src/theme/git_symbol.rs:
```diff
@@ -0,0 +1,52 @@
+impl Default for GitSymbolTheme {
+    fn default() -> GitSymbolTheme {
+        GitSymbolTheme {
+            default: "_".into(),
```
It would be better to differ the `default` and `unmodified` to distinguish them

hpwxf replied:
Now default is a simple space ` `

zwpaper commented on src/theme/git_symbol.rs:
```diff
@@ -0,0 +1,52 @@
+            deleted: "\u{f014}".into(), // or f06
+            modified: "\u{f8ea}".into(),
+            renamed: "\u{f8ea}".into(),
+            ignored: "!".into(),
```
I would prefer a less eye catch symbol for `ignore`, like `I` in `exa`

hpwxf replied:
Done.

zwpaper commented on src/git_symbol.rs:
```diff
@@ -0,0 +1,38 @@
+
+        Self {
+            theme: git_theme,
+            separator: icon_separator,
```
I would prefer not to add this separator here, it looks better without it:

![image](https://user-images.githubusercontent.com/3764335/230782524-00a21ad3-6f08-415c-bb04-8cf91de0bbe1.png)

hpwxf replied:
Done. I also removed the separator option (always no space).

zwpaper reviewed (changes requested):
hi @hpwxf, sorry again for the late reply, I was swamped recently, I had added more comments and it would be ok to go after we resolve them.

thanks so much for the PR, it looks great and we only have some nit to discuss.

let's make the magic happen!

zwpaper commented:
hi @hpwxf, do you have some cycle for this PR, we are planning for 1.0 release, we can try to make it ship with the 1.0

hpwxf commented:
> hi @hpwxf, do you have some cycle for this PR, we are planning for 1.0 release, we can try to make it ship with the 1.0

Sorry for my discontinued availability. I will do the requested changes today or tomorrow.

hpwxf commented:
Thx for the review. Everything solved (see comments for details)

zwpaper commented:
hi @hpwxf, thanks for the great work again, I have only two more nit questions

1. the style is different in `unicode` and `default`
   <img width="679" alt="image" src="https://user-images.githubusercontent.com/3764335/234649282-3d879cd3-a4eb-44e8-bbd8-f14050a45904.png">
2. the space may cause confusion when most of the files are unchanged, but actually I like the clean spaces personally 🤣
   <img width="678" alt="image" src="https://user-images.githubusercontent.com/3764335/234650287-af9b0899-2c31-4e67-ae4f-065e0182b709.png">

Maybe a more consistent choice for the default would be better.

how about a dash `-` for default, and dot `.` for unmodified?

<img width="683" alt="image" src="https://user-images.githubusercontent.com/3764335/234650950-061f3665-2b8a-4bde-8cf9-f9e2c42be689.png">

hpwxf commented:
@zwpaper Thank you for your feedback.
Do you prefer ASCII chars in any mode?
Without the space separator, the symbols are sometimes too close:
<img width="687" alt="Capture d’écran 2023-04-26 à 19 33 11" src="https://user-images.githubusercontent.com/26146722/234656734-e6fa884a-5f2c-4be5-9c1a-b81dcb233301.png">
Never this issue with simple ASCII chars:
<img width="695" alt="Capture d’écran 2023-04-26 à 19 33 40" src="https://user-images.githubusercontent.com/26146722/234656843-1de4ab05-a523-4c8b-985a-c0842060b54d.png">

zwpaper commented:
hi @hpwxf, I would prefer icons, or Unicode, actually, but as you said that it may be too close in some cases, that would be a problem, and it would be too far away if a space added🤣

how about we use ASCII in this first version, and try to figure out a theme that fits all cases?

and we can also let users create a theme file to custom it themselves later.

hpwxf commented:
Hi @zwpaper ;
An idea seems to be:
* default symbols using ASCII chars
* custom symbols using a theme (the user will be able to add a space for large symbol)

I will work on that. (I will first update to only use ASCII chars and then about the theme; if you prefer the git theme can be done in a next PR).

hpwxf commented:
Git theme done:

In config file, you can define:
```yaml
# == Git ==
git:
  # Whether to display git status
  # Possible values: false, true
  enabled: false
  # How to display git status
  # When "classic" is set, this is set to "default".
  # Possible values: default, <theme-file-name>
  # when specifying <theme-file-name>, lsd will look up theme file
  # XDG Base Directory if relative, e.g. ~/.config/lsd/themes/<theme-file-name>.yaml,
  # The file path if absolute
  theme: git
```

and `git.yaml` can be:
```yaml
default: "-"
unmodified: "."
new-in-index: 
new-in-workdir: "?"
deleted: D
modified: ✍
renamed: R
ignored: I
typechange: T
conflicted: C
```

I still have a small bug that always enable git when `--config-file` is used (even if `enabled: false`)

zwpaper commented:
Hi, @hpwxf thanks so much for implementing the theme in this short time!

But I prefer a separate PR for the theme features.

One reason is that we would have more discussions about it like, we were planning to use a fixed theme file name, theme should be optional so that user doesn't have to offer all items, eg., keeping it simple would let us release the 1.0 in a shorter time.
Another one is that I am planning to squash the PR as this is one feature and tens of commits may be a nightmare for others to rebase onto.

zwpaper commented:
Ps it would also be great if we make the git theme (in a separate PR) in time to release with the 1.0

hpwxf commented:
Ok @zwpaper ; all done.

* First I finalize a functional code about a theme for git (only symbols; it could be only with colors)
* Then I did a commit to remove the git theme and I saved the previous work in another branch for a next PR
* I squashed all the commits before your latest review (the latest commits are not yet squashed, but one reviewed it could be a good thing for sure)

tummetott commented:
@hpwxf any news when the git-theme branch will be merged? Is there an issue which tracks the announced discussion around the git-theme?
